### PR TITLE
#77 electric power facility model

### PR DESCRIPTION
**Pull request by nunezco2**
Modifications are in for electric power facility restoration

Result files are provided below
[percentage_of_functionality_wf_restoration.csv](https://github.com/IN-CORE/pyincore/files/7954174/percentage_of_functionality_wf_restoration.csv)
[reptime_wf_restoration.csv](https://github.com/IN-CORE/pyincore/files/7954175/reptime_wf_restoration.csv)

### CODE REVIEW

File: pyincore/dfr3service.py [hunk 1]
```diff
@@ -364,6 +364,70 @@ def _eval_criterion(rule, properties):
...
+        Returns:
+            inventory_class (str): extracted inventory class name. "/" stands for or and "+" stands for and
+
+        """
+        if rules == [[]] or rules == [] or rules == [None]:
+            return "NA"
+        else:
+            inventory_class = ""
+            for i, and_rules in enumerate(rules):
+                # or
```
Comment (diegoac2) [inline pyincore/dfr3service.py:386]: unnecessary comment

File: pyincore/dfr3service.py [hunk 2]
```diff
@@ -364,6 +364,70 @@ def _eval_criterion(rule, properties):
...
+                if i != 0:
+                    inventory_class += "/"
+
+                for j, rule in enumerate(and_rules):
+                    # and
```
Comment (diegoac2) [inline pyincore/dfr3service.py:391]: unnecessary comment

File: pyincore/dfr3service.py [hunk 3]
```diff
@@ -364,6 +364,70 @@ def _eval_criterion(rule, properties):
...
+        """
+
+        if rules == {}:
+            return "NA"
+        else:
+            inventory_class = []
+            boolean = list(rules.keys())[0]  # AND or OR
+            criteria = rules[boolean]
+            for criterion in criteria:
+                # Recursively parse and evaluate the rules with boolean
```
Comment (diegoac2) [inline pyincore/dfr3service.py:415]: what does this mean?

File: pyincore/dfr3service.py [hunk 4]
```diff
@@ -364,6 +364,70 @@ def _eval_criterion(rule, properties):
...
+                if isinstance(criterion, dict):
+                    inventory_class.append(Dfr3Service.extract_inventory_class(criterion))
+                # Base case: get the rule value
```
Comment (diegoac2) [inline pyincore/dfr3service.py:418]: what does this mean?

**Review by diegoac2 — COMMENTED** [review]
added the same comments as water facility restoration

File: pyincore/analyses/electricpowerfacilityrestoration/electricpowerfacilityrestoration.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        Returns:
+            time_results (list): Given Percentage of functionality, the change of repair time
+            pf_results (list): Given Repair time, change of the percentage of functionality
+        """
+
+        time_results = []
+        pf_results = []
+
+        for mapping in mapping_set.mappings:
+            # parse rules to get inventory class. e.g. treatment plan, tank, pump etc
```
Comment (longshuicy) [inline pyincore/analyses/electricpowerfacilityrestoration/electricpowerfacilityrestoration.py:76]: comment only applies to water facility. Maybe you can just remove it here

File: pyincore/analyses/electricpowerfacilityrestoration/electricpowerfacilityrestoration.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                    'type': ['incore:dfr3MappingSet'],
+                }
+            ],
+            'output_datasets': [
+                {
+                    'id': 'pf_results',
+                    'parent_type': '',
+                    'description': 'A csv file recording functionality change with time for each class and limit '
+                                   'state.',
+                    'type': 'incore:electricPowerFacilityRestorationFunc'
```
Comment (longshuicy) [inline pyincore/analyses/electricpowerfacilityrestoration/electricpowerfacilityrestoration.py:164]: you can shorten this to epfRestorationFunc if you want to

File: tests/pyincore/analyses/electricpowerfacilityrestoration/test_electricpowerfacilityrestoration.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+from pyincore import IncoreClient, MappingSet, RestorationCurveSet, RestorationService
+from pyincore.analyses.electricpowerfacilityrestoration import ElectricPowerFacilityRestoration
+import pyincore.globals as pyglobals
+from pyincore.models.mapping import Mapping
+
+
+def run_with_base_class():
+    client = IncoreClient(pyglobals.INCORE_API_DEV_URL)
+    wf_rest = ElectricPowerFacilityRestoration(client)
```
Comment (longshuicy) [inline tests/pyincore/analyses/electricpowerfacilityrestoration/test_electricpowerfacilityrestoration.py:14]: can you rename the variable to epf_rest to avoid confusion?

File: tests/pyincore/analyses/electricpowerfacilityrestoration/test_electricpowerfacilityrestoration.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    restorationsvc = RestorationService(client)
+    mapping_set = MappingSet(restorationsvc.get_mapping("61f302e6e3a03e465500b3eb"))  # new format of mapping
+    wf_rest.set_input_dataset('dfr3_mapping_set', mapping_set)
+    wf_rest.set_parameter("result_name", "wf_restoration.csv")
```
Comment (longshuicy) [inline tests/pyincore/analyses/electricpowerfacilityrestoration/test_electricpowerfacilityrestoration.py:18]: result name shouldn't include .csv. Also could you again rename it to epf_restoration to avoid confusion

**Review by longshuicy — APPROVED** [review]

**Review by diegoac2 — APPROVED** [review]
